activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -115,9 +116,11 @@ def exec_update(sql, name, binds)
       # If the next id was calculated in advance (as in Oracle), it should be
       # passed in as +id_value+.
       def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, binds = [])
-        insert_sql(to_sql(arel, binds), name, pk, id_value, sequence_name, binds)
+        value = exec_insert(to_sql(arel, binds), name, binds, pk, id_value, sequence_name)
+        id_value || last_inserted_id(value)
```

Comment:
I changed the implementation in the PR. `exec_insert` includes `sql_for_insert`, but still returns result object.